File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.cpp
#include <string>
#include <vector>
#include <map>

using namespace std;

int score[8] = {0, 3, 2, 1, 0, 1, 2, 3};

string solution(vector<string> survey, vector<int> choices) {
    string answer = "";
    map<char, int> surv_elm;

    for(int i =0; i<survey.size(); i++)
        surv_elm[survey[i][choices[i]/4]] += score[choices[i]];

    answer += surv_elm['R'] >= surv_elm['T'] ? "R" : "T";
    answer += surv_elm['C'] >= surv_elm['F'] ? "C" : "F";
    answer += surv_elm['J'] >= surv_elm['M'] ? "J" : "M";
    answer += surv_elm['A'] >= surv_elm['N'] ? "A" : "N";

    return answer;
}

File: 프로그래머스/1/118666. 성격 유형 검사하기/성격 유형 검사하기.java
import java.util.*;
import java.util.stream.*;

class Solution {
    public String solution(String[] survey, int[] choices) {
        String answer = "RTCFJMAN";
        
        Map<Character, Integer> map = new HashMap<>();
        
        for(char c : answer.toCharArray())
            map.put(c, 0);
        
        IntStream.range(0, survey.length)
            .forEach(i -> {
                
                int choice = choices[i];
                char cur = ' ';
                if(choice > 4){
                    cur = survey[i].charAt(1);
                }else{
                    cur = survey[i].charAt(0);
                }
                map.put(cur, map.getOrDefault(cur, 0) + 
                       Math.abs(choice - 4));
            });
        
        
        StringBuilder result = new StringBuilder();
        for(int i = 0 ; i < answer.length() ; i += 2){
            result.append(
            map.get(answer.charAt(i)) >= map.get(answer.charAt(i + 1)) ?
                answer.charAt(i) : answer.charAt(i + 1)
            );
        }
        
        return result.toString();
    }
}